File: flapypan-blog-api/src/main/java/top/flapypan/blog/service/SettingService.java
package top.flapypan.blog.service;

import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import top.flapypan.blog.entity.Setting;
import top.flapypan.blog.repository.SettingRepository;

import java.util.*;
import java.util.stream.Collectors;

@Slf4j
@RequiredArgsConstructor
@Service
public class SettingService {

    // 默认设置
    public static final Map<String, String> DEFAULT_SETTINGS = Map.of(
            "siteTitle", "FlapyPan's Blog",
            "favicon", "/avatar.webp",
            "avatar", "/avatar.webp",
            "banner", "/banner.webp",
            "name", "FlapyPan",
            "email", "flapypan@gmail.com",
            "info", "个人博客",
            "pageSize", "12",
            "footer", "Copyright"
    );

    private final SettingRepository repository;

    public void initDefaultSettings() {
        int defaultSize = DEFAULT_SETTINGS.size();
        if (repository.count() >= defaultSize) return;
        synchronized (SettingService.class) {
            // 获取已经有的key
            Set<String> allKey = repository.findAllKey();
            // 求差集
            Set<String> keysToAdd = new HashSet<>(DEFAULT_SETTINGS.keySet());
            keysToAdd.removeAll(allKey);
            // 生成需要插入的值
            List<Setting> settingsToAdd = keysToAdd.stream()
                    .map(key -> new Setting(key, DEFAULT_SETTINGS.get(key)))
                    .toList();
            int size = repository.saveAllAndFlush(settingsToAdd).size();
            log.info("添加了 {} 项默认设置", size);
        }
    }

    public Map<String, String> getSettingsMap() {
        initDefaultSettings();
        return repository
                .findAll()
                .stream()
                .collect(Collectors.toMap(Setting::getKey, Setting::getValue));
    }

    @Transactional
    public void saveSettingsMap(Map<String, String> settingsMap) {
        List<Setting> settingsWillSaved = repository.findAll().stream() // 获取所有设置
                .filter(s -> settingsMap.containsKey(s.getKey())) // 过滤掉无需存储的
                .filter(s -> !Objects.equals(settingsMap.get(s.getKey()), s.getValue())) // 过滤掉值相等的
                .peek(s -> s.setValue(settingsMap.get(s.getKey()))) // 设置新值
                .toList();
        repository.saveAllAndFlush(settingsWillSaved);
    }

}

File: flapypan-blog-api/src/main/kotlin/top/flapypan/blog/service/SettingService.kt
package top.flapypan.blog.service

import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import top.flapypan.blog.common.LoggerDelegate
import top.flapypan.blog.entity.Setting
import top.flapypan.blog.repository.SettingRepository

// 默认设置
private val DEFAULT_SETTINGS = mapOf(
    "siteTitle" to "FlapyPan's Blog",
    "favicon" to "/avatar.webp",
    "avatar" to "/avatar.webp",
    "banner" to "/banner.webp",
    "name" to "FlapyPan",
    "email" to "flapypan@gmail.com",
    "info" to "个人博客",
    "pageSize" to "12",
    "footer" to "Copyright",
    "giscusRepo" to "",
    "giscusRepoId" to "",
    "giscusCategory" to "",
    "giscusCategoryId" to "",
)

@Service
class SettingService(
    private val repository: SettingRepository
) {

    private val log by LoggerDelegate()

    fun initDefaultSettings() {
        if (repository.count() >= DEFAULT_SETTINGS.size) return
        synchronized(SettingService::class) {
            // 获取已经有的key
            val allKey = repository.findAllKey()
            // 求差集
            val keysToAdd = DEFAULT_SETTINGS.keys - allKey
            // 生成需要插入的值
            val settingsToAdd = keysToAdd.map { key -> Setting(key, DEFAULT_SETTINGS[key]) }
            val size = repository.saveAllAndFlush(settingsToAdd).size
            log.info("添加了 $size 项默认设置")
        }
    }

    fun getSettingsMap(): Map<String, String?> {
        initDefaultSettings()
        return repository.findAll()
            .associateBy(Setting::key, Setting::value)
    }

    @Transactional
    fun saveSettingsMap(settingsMap: Map<String, String?>) {
        // 获取所有设置
        val settingsWillSaved = repository.findAll()
            // 过滤掉无需存储的t和值相等的
            .filter { it.key in settingsMap && settingsMap[it.key] != it.value }
            // 设置新值
            .onEach { it.value = settingsMap[it.key] }
        repository.saveAllAndFlush(settingsWillSaved)
    }

}
